
# ** Business Understanding **

#### **Problem Statement - Foodhub**

##### ** Context**
The number of restaurants in New York is increasing day by day. Lots of students and busy professionals rely on those restaurants due to their hectic lifestyles. Online food delivery service is a great option for them. It provides them with good food from their favorite restaurants. A food aggregator company FoodHub offers access to multiple restaurants through a single smartphone app.

The app allows the restaurants to receive a direct online order from a customer. The app assigns a delivery person from the company to pick up the order after it is confirmed by the restaurant. The delivery person then uses the map to reach the restaurant and waits for the food package. Once the food package is handed over to the delivery person, he/she confirms the pick-up in the app and travels to the customer's location to deliver the food. The delivery person confirms the drop-off in the app after delivering the food package to the customer. The customer can rate the order in the app. The food aggregator earns money by collecting a fixed margin of the delivery order from the restaurants.



## ** Objective ***
The food aggregator company has stored the data of the different orders made by the registered customers in their online portal. They want to analyze the data to get a fair idea about the demand of different restaurants which will help them in enhancing their customer experience. Suppose you are hired as a Data Scientist in this company and the Data Science team has shared some of the key questions that need to be answered. Perform the data analysis to find answers to these questions that will help the company to improve the business.

### ** Foodhub Data Description **
The data contains the different data related to a food order. The detailed data dictionary is given below.
Data Dictionary
    - order_id: Unique ID of the order
    - customer_id: ID of the customer who ordered the food
    - restaurant_name: Name of the restaurant
    - cuisine_type: Cuisine ordered by the customer
    - cost: Cost of the order
    - day_of_the_week: Indicates whether the order is placed on a weekday or weekend (The weekday is from Monday to Friday and the weekend is Saturday and Sunday)
    - rating: Rating given by the customer out of 5
    - food_preparation_time: Time (in minutes) taken by the restaurant to prepare the food. This is calculated by taking the difference between the timestamps of the restaurant's order confirmation and the delivery person's pick-up confirmation.
    - delivery_time: Time (in minutes) taken by the delivery person to deliver the food package. This is calculated by taking the difference between the timestamps of the delivery person's pick-up confirmation and drop-off information

#### ** Data Understanding **

- head and tail (printing out the first 5 or few defined entries and the last entries)
- shape (dimensions)
- columns
- datatypes
- info
- describe stats numerical  - mean | median | mode
- describe stats categorical  - count | percentages


##### Load required libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot  as plt

pd.set_option('display.max_columns', None)  # Show all columns   

#### Load data

In [3]:
foodhub_df = pd.read_csv("data/foodhub_order.csv") 

#### Get a glimpse of the foodhub dataset

In [ ]:
foodhub_df.head() # First five rows of the foodhub dataset

,order_id,customer_id,restaurant_name,cuisine_type,cost_of_the_order,day_of_the_week,rating,food_preparation_time,delivery_time
0,1477147,337525,Hangawi,Korean,30.75,Weekend,Not given,25,20
1,1477685,358141,Blue Ribbon Sushi Izakaya,Japanese,12.08,Weekend,Not given,25,23
2,1477070,66393,Cafe Habana,Mexican,12.23,Weekday,5,23,28
3,1477334,106968,Blue Ribbon Fried Chicken,American,29.20,Weekend,3,25,15
4,1478249,76942,Dirty Bird to Go,American,11.59,Weekday,4,25,24


In [ ]:
foodhub_df.head(10) # First ten rows of the foodhub dataset

,order_id,customer_id,restaurant_name,cuisine_type,cost_of_the_order,day_of_the_week,rating,food_preparation_time,delivery_time
0,1477147,337525,Hangawi,Korean,30.75,Weekend,Not given,25,20
1,1477685,358141,Blue Ribbon Sushi Izakaya,Japanese,12.08,Weekend,Not given,25,23
2,1477070,66393,Cafe Habana,Mexican,12.23,Weekday,5,23,28
3,1477334,106968,Blue Ribbon Fried Chicken,American,29.20,Weekend,3,25,15
4,1478249,76942,Dirty Bird to Go,American,11.59,Weekday,4,25,24
5,1477224,147468,Tamarind TriBeCa,Indian,25.22,Weekday,3,20,24
6,1477894,157711,The Meatball Shop,Italian,6.07,Weekend,Not given,28,21
7,1477859,89574,Barbounia,Mediterranean,5.97,Weekday,3,33,30
8,1477174,121706,Anjappar Chettinad,Indian,16.44,Weekday,5,21,26
9,1477311,39705,Bukhara Grill,Indian,7.18,Weekday,5,29,26


In [6]:
foodhub_df.tail() # Last five rows of the foodhub dataset

,order_id,customer_id,restaurant_name,cuisine_type,cost_of_the_order,day_of_the_week,rating,food_preparation_time,delivery_time
1893,1476701,292602,Chipotle Mexican Grill $1.99 Delivery,Mexican,22.31,Weekend,5,31,17
1894,1477421,397537,The Smile,American,12.18,Weekend,5,31,19
1895,1477819,35309,Blue Ribbon Sushi,Japanese,25.22,Weekday,Not given,31,24
1896,1477513,64151,Jack's Wife Freda,Mediterranean,12.18,Weekday,5,23,31
1897,1478056,120353,Blue Ribbon Sushi,Japanese,19.45,Weekend,Not given,28,24


In [7]:
foodhub_df.tail(10) # Last ten rows of the foodhub dataset

,order_id,customer_id,restaurant_name,cuisine_type,cost_of_the_order,day_of_the_week,rating,food_preparation_time,delivery_time
1888,1477353,106324,The Meatball Shop,Italian,16.20,Weekend,5,21,26
1889,1478190,94152,RedFarm Broadway,Chinese,8.68,Weekday,3,33,30
1890,1477316,164776,TAO,Japanese,15.67,Weekend,5,20,22
1891,1476981,138586,Shake Shack,American,5.82,Weekend,Not given,22,28
1892,1477473,97838,Han Dynasty,Chinese,29.15,Weekend,Not given,29,21
1893,1476701,292602,Chipotle Mexican Grill $1.99 Delivery,Mexican,22.31,Weekend,5,31,17
1894,1477421,397537,The Smile,American,12.18,Weekend,5,31,19
1895,1477819,35309,Blue Ribbon Sushi,Japanese,25.22,Weekday,Not given,31,24
1896,1477513,64151,Jack's Wife Freda,Mediterranean,12.18,Weekday,5,23,31
1897,1478056,120353,Blue Ribbon Sushi,Japanese,19.45,Weekend,Not given,28,24


In [14]:
foodhub_df.isnull().sum() # cheking if therer are any missing values

order_id                 0
customer_id              0
restaurant_name          0
cuisine_type             0
cost_of_the_order        0
day_of_the_week          0
rating                   0
food_preparation_time    0
delivery_time            0
dtype: int64

Insight of the dataset: 
- Dataset has no missing values

In [ ]:
# shape (dimensions) of the foodhub dataset
rows, columns =  foodhub_df.shape 
output = f"ROWS: {rows} || COLUMNS: {columns}"
print(output)

ROWS: 1898 || COLUMNS: 9


Insight of the dataset: 
- Dataset has 1898 records and 9 columns

In [10]:
foodhub_df.info() # displays info about the foodhub dataset

<class 'pandas.DataFrame'>
RangeIndex: 1898 entries, 0 to 1897
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               1898 non-null   int64  
 1   customer_id            1898 non-null   int64  
 2   restaurant_name        1898 non-null   str    
 3   cuisine_type           1898 non-null   str    
 4   cost_of_the_order      1898 non-null   float64
 5   day_of_the_week        1898 non-null   str    
 6   rating                 1898 non-null   str    
 7   food_preparation_time  1898 non-null   int64  
 8   delivery_time          1898 non-null   int64  
dtypes: float64(1), int64(4), str(4)
memory usage: 133.6 KB


In [ ]:
foodhub_df.dtypes # displays data types present in the foodhub dataset

order_id                   int64
customer_id                int64
restaurant_name              str
cuisine_type                 str
cost_of_the_order        float64
day_of_the_week              str
rating                       str
food_preparation_time      int64
delivery_time              int64
dtype: object

Insights: 
- composed of  both categorical and numerical data in form of 64bit integers for whole numbers, string data type and 64bit float for decimals

##### Descriptive Stats

- MCTs -> measures of central tendency (Mean, Median, Mode)
- MoV -> measures of variation (range, IQR, percentile, quartiles,variance , standard deviation )

In [12]:
foodhub_df.describe() # get desciptive stats of numerical data

,order_id,customer_id,cost_of_the_order,food_preparation_time,delivery_time
count,1.898000e+03,1898.000000,1898.000000,1898.000000,1898.000000
mean,1.477496e+06,171168.478398,16.498851,27.371970,24.161749
std,5.480497e+02,113698.139743,7.483812,4.632481,4.972637
min,1.476547e+06,1311.000000,4.470000,20.000000,15.000000
25%,1.477021e+06,77787.750000,12.080000,23.000000,20.000000
50%,1.477496e+06,128600.000000,14.140000,27.000000,25.000000
75%,1.477970e+06,270525.000000,22.297500,31.000000,28.000000
max,1.478444e+06,405334.000000,35.410000,35.000000,33.000000


Insights:

1. Order Cost:
    - Orders cost is between a minimum of 4.47 and a maximum of 35.41
    - 75% of the costs are above 22.29
2. Food Preparation Time:
    - Time for food preparation is between a minimum of 20 minutes and a maximum of 35 minutes
    - 75% of the orsers had an average preparation time of 31 minutes
3. Delivery Time:
    - Time taken to deliver orders ranged from a minimum of 15 minutes and a maximum of 33 minutes
    - 75% of the orders had an average delivery time of 28 minutes

In [ ]:
foodhub_df.describe(include='object') 

/var/folders/0c/1j_h2q_d3cz5xx15_x7nbdyw0000gn/T/ipykernel_73101/3868679781.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  foodhub_df.describe(include='object') # get desciptive stats of numerical data


,restaurant_name,cuisine_type,day_of_the_week,rating
count,1898,1898,1898,1898
unique,178,14,2,4
top,Shake Shack,American,Weekend,Not given
freq,219,584,1351,736


Insights:
- Top restaurant was Shake shack with a frequency of 219
- Top cuisine type ordered was American with a frequency of 584
- Most orders were received duting the weekend with a frequency of 1,351
- 736 orders placed were not rated

In [29]:
foodhub_df.describe() # get desciptive stats of categorical data
print(foodhub_df)

      order_id  customer_id                        restaurant_name  \
0      1477147       337525                                Hangawi   
1      1477685       358141              Blue Ribbon Sushi Izakaya   
2      1477070        66393                            Cafe Habana   
3      1477334       106968              Blue Ribbon Fried Chicken   
4      1478249        76942                       Dirty Bird to Go   
...        ...          ...                                    ...   
1893   1476701       292602  Chipotle Mexican Grill $1.99 Delivery   
1894   1477421       397537                              The Smile   
1895   1477819        35309                      Blue Ribbon Sushi   
1896   1477513        64151                      Jack's Wife Freda   
1897   1478056       120353                      Blue Ribbon Sushi   

       cuisine_type  cost_of_the_order day_of_the_week     rating  \
0            Korean              30.75         Weekend  Not given   
1          Japanese  

In [30]:
foodhub_df[["restaurant_name"]].describe()

,restaurant_name
count,1898
unique,178
top,Shake Shack
freq,219


Insights:
- Top restaurant was Shake shack with a frequency of 219

In [32]:
foodhub_df[["cuisine_type"]].describe()

,cuisine_type
count,1898
unique,14
top,American
freq,584


Insights:
- Top cuisine type ordered was American with a frequency of 584

In [33]:
foodhub_df[["day_of_the_week"]].describe()

,day_of_the_week
count,1898
unique,2
top,Weekend
freq,1351


Insights:
- Most orders were received duting the weekend with a frequency of 1,351

In [34]:
foodhub_df[["rating"]].describe()

,rating
count,1898
unique,4
top,Not given
freq,736


Insights:
- 736 orders placed were not rated

In [35]:
foodhub_df[["food_preparation_time"]].describe()

,food_preparation_time
count,1898.000000
mean,27.371970
std,4.632481
min,20.000000
25%,23.000000
50%,27.000000
75%,31.000000
max,35.000000


Insights:
1. Food Preparation Time:
    - Time for food preparation is between a minimum of 20 minutes and a maximum of 35 minutes
    - 75% of the orsers had an average preparation time of 31 minutes


In [36]:
foodhub_df[["delivery_time"]].describe()

,delivery_time
count,1898.000000
mean,24.161749
std,4.972637
min,15.000000
25%,20.000000
50%,25.000000
75%,28.000000
max,33.000000


Insights:
1. Delivery Time:
    - Time taken to deliver orders ranged from a minimum of 15 minutes and a maximum of 33 minutes
    - 75% of the orders had an average delivery time of 28 minutes